presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java

```diff
@@ -107,6 +111,19 @@
     private final String clientInfo;
     private final Map<SqlFunctionId, SqlInvokedFunction> sessionFunctions;
 
+    private Optional<Tracer> tracer;
+
+    public HttpRequestSessionContext(HttpServletRequest servletRequest, SqlParserOptions sqlParserOptions, Tracer tracer)
+    {
+        this(servletRequest, sqlParserOptions);
+        if (isTracingEnabled()) {
+            this.tracer = Optional.of(requireNonNull(tracer, "tracer is null"));
+        }
+        else {
+            this.tracer = Optional.of(NoopTracerProvider.NOOP_TRACER);
+        }
```

Comment:
This should be inside the other constructor. I understand you wanna pass in tracer as it is created earlier than session. So I can't make too much comment at the moment before seeing the full picture of the complete creation flow? Maybe let's finish the PR and see how we can better store tracer in this class?

Reply:
Sounds good!